## Datasets:

https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv

https://raw.githubusercontent.com/justmarkham/DAT8/master/data/beer.txt

https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/04_Apply/US_Crime_Rates/US_Crime_Rates_1960_2014.csv

https://raw.githubusercontent.com/vincentarelbundock/Rdatasets/master/datasets.csv

https://drive.google.com/file/d/1etGP85wJWsgqbsh3gX5oxc8dPKLMA1eT/view

In [3]:
import pandas as pd, os

# Dataset 1:

In [2]:
ds1=pd.read_table("https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv")
ds1

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

In [3]:
len(ds1)

4622

### Create a column having total price of items i.e. quantity * item_price:

In [5]:
price = [float(ds1['item_price'][i][1:]) for i in range(len(ds1))]

#creating a column that does the needful
ds1['Amount_paid'] = price * ds1['quantity']

#adding $ before the total amount paid:

#creating a list that contains the total amount paid with $ sign.
dollars=[f"${str(ds1['Amount_paid'][i])}" for i in range(len(ds1))]

#implementing the column we desired.
ds1['Amount_paid'] = dollars
ds1

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price Amount_paid  
0                                                   NaN     $2.39        $2.39  
1                                          [Clementine]     $3.39        $3.39  
2                                               [Apple]     $3.39        $3.39  
3                                                   NaN     $2.39        $2.39  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98       $33.96  
...                                                 ...        ...         ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75       $11.75  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75       $11.75  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25       $11.25  
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75        $8.75  
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75        $8.75  

[4622 rows x 6 columns]

## Find out the unique items and the total number of times an item is bought:

### i) Unique Items:

In [6]:
ds1[['item_name']].describe()

item_name
count           4622
unique            50
top     Chicken Bowl
freq             726

In [7]:
#creating an array of unique_items using .unique() and converting it into DataFrame.
unique_items= pd.DataFrame((ds1['item_name'].unique()))

#creating a column 'unique_items' and assigning it the values of the array created.
ds1[['items_offered']] = unique_items

ds1

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  \
0                                                   NaN     $2.39    
1                                          [Clementine]     $3.39    
2                                               [Apple]     $3.39    
3                                                   NaN     $2.39    
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98    
...                                                 ...        ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75    
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75    
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25    
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75    
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75    

     Amount_paid                          items_offered  
0          $2.39           Chips and Fresh Tomato Salsa  
1          $3.39                                   Izze  
2          $3.39                       Nantucket Nectar  
3          $2.39  Chips and Tomatillo-Green Chili Salsa  
4         $33.96                           Chicken Bowl  
...          ...                                    ...  
4617      $11.75                                    NaN  
4618      $11.75                                    NaN  
4619      $11.25                                    NaN  
4620       $8.75                                    NaN  
4621       $8.75                                    NaN  

[4622 rows x 7 columns]

In [9]:
ds1[['items_offered']][ds1['items_offered'].isnull()==False]

items_offered
0            Chips and Fresh Tomato Salsa
1                                    Izze
2                        Nantucket Nectar
3   Chips and Tomatillo-Green Chili Salsa
4                            Chicken Bowl
5                           Side of Chips
6                           Steak Burrito
7                        Steak Soft Tacos
8                     Chips and Guacamole
9                    Chicken Crispy Tacos
10                     Chicken Soft Tacos
11                        Chicken Burrito
12                            Canned Soda
13                       Barbacoa Burrito
14                       Carnitas Burrito
15                          Carnitas Bowl
16                          Bottled Water
17  Chips and Tomatillo Green Chili Salsa
18                          Barbacoa Bowl
19                                  Chips
20                     Chicken Salad Bowl
21                             Steak Bowl
22                    Barbacoa Soft Tacos
23                         Veggie Burrito
24                            Veggie Bowl
25                     Steak Crispy Tacos
26    Chips and Tomatillo Red Chili Salsa
27                  Barbacoa Crispy Tacos
28                      Veggie Salad Bowl
29     Chips and Roasted Chili-Corn Salsa
30     Chips and Roasted Chili Corn Salsa
31                    Carnitas Soft Tacos
32                          Chicken Salad
33                      Canned Soft Drink
34                       Steak Salad Bowl
35                      6 Pack Soft Drink
36    Chips and Tomatillo-Red Chili Salsa
37                                   Bowl
38                                Burrito
39                           Crispy Tacos
40                  Carnitas Crispy Tacos
41                            Steak Salad
42      Chips and Mild Fresh Tomato Salsa
43                      Veggie Soft Tacos
44                    Carnitas Salad Bowl
45                    Barbacoa Salad Bowl
46                                  Salad
47                    Veggie Crispy Tacos
48                           Veggie Salad
49                         Carnitas Salad

### ii) total number of times an item is bought:

In [6]:
ds1.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

In [17]:
ds1.groupby('item_name').sum().sort_values(by='item_name',ascending=True)[['quantity']]

quantity
item_name                                      
6 Pack Soft Drink                            55
Barbacoa Bowl                                66
Barbacoa Burrito                             91
Barbacoa Crispy Tacos                        12
Barbacoa Salad Bowl                          10
Barbacoa Soft Tacos                          25
Bottled Water                               211
Bowl                                          4
Burrito                                       6
Canned Soda                                 126
Canned Soft Drink                           351
Carnitas Bowl                                71
Carnitas Burrito                             60
Carnitas Crispy Tacos                         8
Carnitas Salad                                1
Carnitas Salad Bowl                           6
Carnitas Soft Tacos                          40
Chicken Bowl                                761
Chicken Burrito                             591
Chicken Crispy Tacos                         50
Chicken Salad                                 9
Chicken Salad Bowl                          123
Chicken Soft Tacos                          120
Chips                                       230
Chips and Fresh Tomato Salsa                130
Chips and Guacamole                         506
Chips and Mild Fresh Tomato Salsa             1
Chips and Roasted Chili Corn Salsa           23
Chips and Roasted Chili-Corn Salsa           18
Chips and Tomatillo Green Chili Salsa        45
Chips and Tomatillo Red Chili Salsa          50
Chips and Tomatillo-Green Chili Salsa        33
Chips and Tomatillo-Red Chili Salsa          25
Crispy Tacos                                  2
Izze                                         20
Nantucket Nectar                             29
Salad                                         2
Side of Chips                               110
Steak Bowl                                  221
Steak Burrito                               386
Steak Crispy Tacos                           36
Steak Salad                                   4
Steak Salad Bowl                             31
Steak Soft Tacos                             56
Veggie Bowl                                  87
Veggie Burrito                               97
Veggie Crispy Tacos                           1
Veggie Salad                                  6
Veggie Salad Bowl                            18
Veggie Soft Tacos                             8

### .value_counts(): To count the number of occurences.

In [16]:
ds1[['item_name']].value_counts().sort_values(ascending=True)

item_name                            
Veggie Crispy Tacos                        1
Carnitas Salad                             1
Chips and Mild Fresh Tomato Salsa          1
Crispy Tacos                               2
Salad                                      2
Bowl                                       2
Steak Salad                                4
Carnitas Salad Bowl                        6
Veggie Salad                               6
Burrito                                    6
Veggie Soft Tacos                          7
Carnitas Crispy Tacos                      7
Chicken Salad                              9
Barbacoa Salad Bowl                       10
Barbacoa Crispy Tacos                     11
Chips and Roasted Chili-Corn Salsa        18
Veggie Salad Bowl                         18
Izze                                      20
Chips and Tomatillo-Red Chili Salsa       20
Chips and Roasted Chili Corn Salsa        22
Barbacoa Soft Tacos                       25
Nantucket Nectar 

In [5]:
ds1['item_name'].unique()

array(['Chips and Fresh Tomato Salsa', 'Izze', 'Nantucket Nectar',
       'Chips and Tomatillo-Green Chili Salsa', 'Chicken Bowl',
       'Side of Chips', 'Steak Burrito', 'Steak Soft Tacos',
       'Chips and Guacamole', 'Chicken Crispy Tacos',
       'Chicken Soft Tacos', 'Chicken Burrito', 'Canned Soda',
       'Barbacoa Burrito', 'Carnitas Burrito', 'Carnitas Bowl',
       'Bottled Water', 'Chips and Tomatillo Green Chili Salsa',
       'Barbacoa Bowl', 'Chips', 'Chicken Salad Bowl', 'Steak Bowl',
       'Barbacoa Soft Tacos', 'Veggie Burrito', 'Veggie Bowl',
       'Steak Crispy Tacos', 'Chips and Tomatillo Red Chili Salsa',
       'Barbacoa Crispy Tacos', 'Veggie Salad Bowl',
       'Chips and Roasted Chili-Corn Salsa',
       'Chips and Roasted Chili Corn Salsa', 'Carnitas Soft Tacos',
       'Chicken Salad', 'Canned Soft Drink', 'Steak Salad Bowl',
       '6 Pack Soft Drink', 'Chips and Tomatillo-Red Chili Salsa', 'Bowl',
       'Burrito', 'Crispy Tacos', 'Carnitas Crispy Tacos

In [12]:
ds1[['quantity']][ds1['item_name']=="Chips and Fresh Tomato Salsa"]

quantity
0            1
25           1
55           1
89           1
183          1
...        ...
4231         1
4318         1
4324         1
4425         1
4503         1

[110 rows x 1 columns]

In [20]:
x=ds1[['quantity']][ds1['item_name']=="Chips and Fresh Tomato Salsa"].sum()
x

quantity    130
dtype: int64

In [23]:
#Creating a list that will contain the no. of times an item is bought.
timesbought = []

#filling in the values.
for i in ds1['items_offered'][ds1['items_offered'].isnull() == False]:
    timesbought.append(ds1[['quantity']][ds1['item_name']==i].sum())

#Converting the list into DataFrame datatype. 
timesbought = pd.DataFrame(timesbought)

#Creating a column we desired and assigning it the values we've had in the list. 
ds1[["No. of times an item is bought"]] = timesbought

ds1.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price Amount_paid  \
0                                                NaN     $2.39        $2.39   
1                                       [Clementine]     $3.39        $3.39   
2                                            [Apple]     $3.39        $3.39   
3                                                NaN     $2.39        $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98       $33.96   

                           items_offered  No. of times an item is bought  
0           Chips and Fresh Tomato Salsa                           130.0  
1                                   Izze                            20.0  
2                       Nantucket Nectar                            29.0  
3  Chips and Tomatillo-Green Chili Salsa                            33.0  
4                           Chicken Bowl                           761.0

## Results Table:

In [16]:
res=ds1[['items_offered']][ds1['items_offered'].isnull()==False]
res['No. of times an item is bought'] = ds1['No. of times an item is bought'][ds1["No. of times an item is bought"].isnull()==False]
res

items_offered  No. of times an item is bought
0            Chips and Fresh Tomato Salsa                           130.0
1                                    Izze                            20.0
2                        Nantucket Nectar                            29.0
3   Chips and Tomatillo-Green Chili Salsa                            33.0
4                            Chicken Bowl                           761.0
5                           Side of Chips                           110.0
6                           Steak Burrito                           386.0
7                        Steak Soft Tacos                            56.0
8                     Chips and Guacamole                           506.0
9                    Chicken Crispy Tacos                            50.0
10                     Chicken Soft Tacos                           120.0
11                        Chicken Burrito                           591.0
12                            Canned Soda                           126.0
13                       Barbacoa Burrito                            91.0
14                       Carnitas Burrito                            60.0
15                          Carnitas Bowl                            71.0
16                          Bottled Water                           211.0
17  Chips and Tomatillo Green Chili Salsa                            45.0
18                          Barbacoa Bowl                            66.0
19                                  Chips                           230.0
20                     Chicken Salad Bowl                           123.0
21                             Steak Bowl                           221.0
22                    Barbacoa Soft Tacos                            25.0
23                         Veggie Burrito                            97.0
24                            Veggie Bowl                            87.0
25                     Steak Crispy Tacos                            36.0
26    Chips and Tomatillo Red Chili Salsa                            50.0
27                  Barbacoa Crispy Tacos                            12.0
28                      Veggie Salad Bowl                            18.0
29     Chips and Roasted Chili-Corn Salsa                            18.0
30     Chips and Roasted Chili Corn Salsa                            23.0
31                    Carnitas Soft Tacos                            40.0
32                          Chicken Salad                             9.0
33                      Canned Soft Drink                           351.0
34                       Steak Salad Bowl                            31.0
35                      6 Pack Soft Drink                            55.0
36    Chips and Tomatillo-Red Chili Salsa                            25.0
37                                   Bowl                             4.0
38                                Burrito                             6.0
39                           Crispy Tacos                             2.0
40                  Carnitas Crispy Tacos                             8.0
41                            Steak Salad                             4.0
42      Chips and Mild Fresh Tomato Salsa                             1.0
43                      Veggie Soft Tacos                             8.0
44                    Carnitas Salad Bowl                             6.0
45                    Barbacoa Salad Bowl                            10.0
46                                  Salad                             2.0
47                    Veggie Crispy Tacos                             1.0
48                           Veggie Salad                             6.0
49                         Carnitas Salad                             1.0

In [27]:
ds1['choice_description'].describe()

count            3376
unique           1043
top       [Diet Coke]
freq              134
Name: choice_description, dtype: object

In [29]:
pd.Categorical(ds1['choice_description'])

[NaN, '[Clementine]', '[Apple]', NaN, '[Tomatillo-Red Chili Salsa (Hot), [Black Bean..., ..., '[Fresh Tomato Salsa, [Rice, Black Beans, Sour..., '[Fresh Tomato Salsa, [Rice, Sour Cream, Chees..., '[Fresh Tomato Salsa, [Fajita Vegetables, Pint..., '[Fresh Tomato Salsa, [Fajita Vegetables, Lett..., '[Fresh Tomato Salsa, [Fajita Vegetables, Pint...]
Length: 4622
Categories (1043, object): ['[Adobo-Marinated and Grilled Chicken, Pinto B..., '[Adobo-Marinated and Grilled Chicken, [Sour C..., '[Adobo-Marinated and Grilled Chicken]', '[Adobo-Marinated and Grilled Steak, [Sour Cre..., ..., '[[Tomatillo-Red Chili Salsa (Hot), Tomatillo-..., '[[Tomatillo-Red Chili Salsa (Hot), Tomatillo-..., '[[Tomatillo-Red Chili Salsa (Hot), Tomatillo-..., '[[Tomatillo-Red Chili Salsa (Hot), Tomatillo-...]

## .info(): 
 - Prints a **concise summary of a DataFrame**.
 - This method prints information about a DataFrame including the **index dtype** and **columns, non-null values** and **memory usage.**

In [30]:
ds1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 8 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   order_id                        4622 non-null   int64  
 1   quantity                        4622 non-null   int64  
 2   item_name                       4622 non-null   object 
 3   choice_description              3376 non-null   object 
 4   item_price                      4622 non-null   object 
 5   Amount_paid                     4622 non-null   object 
 6   items_offered                   50 non-null     object 
 7   No. of times an item is bought  50 non-null     float64
dtypes: float64(1), int64(2), object(5)
memory usage: 289.0+ KB


In [35]:
ds1.index

RangeIndex(start=0, stop=4622, step=1)

## Which was the most ordered item? and How many items were ordered?

In [42]:
ds1[['item_name']].describe()

item_name
count           4622
unique            50
top     Chicken Bowl
freq             726

**Chicken Bowl was most ordered.**

In [47]:
ds1.groupby('item_name').sum().sort_values(['quantity'], ascending=False).head(1)

order_id  quantity  No. of times an item is bought
item_name                                                       
Chicken Bowl    713926       761                          1243.0

**Chicken bowl was ordered 1243 times.**

## What was the most ordered item in the choice_description column?

In [48]:
ds1.head(5)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price Amount_paid  \
0                                                NaN     $2.39        $2.39   
1                                       [Clementine]     $3.39        $3.39   
2                                            [Apple]     $3.39        $3.39   
3                                                NaN     $2.39        $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98       $33.96   

                           items_offered  No. of times an item is bought  
0           Chips and Fresh Tomato Salsa                           130.0  
1                                   Izze                            20.0  
2                       Nantucket Nectar                            29.0  
3  Chips and Tomatillo-Green Chili Salsa                            33.0  
4                           Chicken Bowl                           761.0

In [52]:
ds1.groupby('choice_description').sum().sort_values(['quantity'], ascending=False).head(1)

order_id  quantity  No. of times an item is bought
choice_description                                                    
[Diet Coke]           123455       159                             0.0

**[Diet Coke] was ordered 159 times.**

## How much was the revenue for the period in the dataset?

In [73]:
#revenue = quantity * price
#or sum of Amount_paid column

#creating a list of elements of 'amount_paid' column having dollar sign removed and data type converted to float.
amount_paid = [ float(ds1['Amount_paid'][i][1:]) for i in range(len(ds1))]
amount_paid = pd.Series(amount_paid)

revenue = amount_paid.sum()
revenue

39237.02

## Print a data frame with only two columns item_name and item_price:

In [98]:
ds1[['item_name', 'item_price']].drop_duplicates(['item_name']).sort_values(['item_name'], ascending=True)

item_name item_price
298                       6 Pack Soft Drink     $6.49 
39                            Barbacoa Bowl    $11.75 
21                         Barbacoa Burrito     $8.99 
168                   Barbacoa Crispy Tacos    $11.75 
1229                    Barbacoa Salad Bowl    $11.89 
56                      Barbacoa Soft Tacos     $9.25 
34                            Bottled Water     $1.09 
445                                    Bowl    $22.20 
510                                 Burrito     $7.40 
18                              Canned Soda     $2.18 
263                       Canned Soft Drink     $1.25 
33                            Carnitas Bowl     $8.99 
27                         Carnitas Burrito     $8.99 
554                   Carnitas Crispy Tacos     $9.25 
3750                         Carnitas Salad     $8.99 
1132                    Carnitas Salad Bowl    $11.89 
237                     Carnitas Soft Tacos     $9.25 
4                              Chicken Bowl    $16.98 
16                          Chicken Burrito     $8.49 
11                     Chicken Crispy Tacos     $8.75 
250                           Chicken Salad    $10.98 
44                       Chicken Salad Bowl     $8.75 
12                       Chicken Soft Tacos     $8.75 
40                                    Chips     $2.15 
0              Chips and Fresh Tomato Salsa     $2.39 
10                      Chips and Guacamole     $4.45 
674       Chips and Mild Fresh Tomato Salsa     $3.00 
233      Chips and Roasted Chili Corn Salsa     $2.95 
191      Chips and Roasted Chili-Corn Salsa     $2.39 
38    Chips and Tomatillo Green Chili Salsa     $2.95 
111     Chips and Tomatillo Red Chili Salsa     $2.95 
3     Chips and Tomatillo-Green Chili Salsa     $2.39 
300     Chips and Tomatillo-Red Chili Salsa     $2.39 
520                            Crispy Tacos     $7.40 
1                                      Izze     $3.39 
2                          Nantucket Nectar     $3.39 
1414                                  Salad     $7.40 
6                             Side of Chips     $1.69 
54                               Steak Bowl     $8.99 
7                             Steak Burrito    $11.75 
92                       Steak Crispy Tacos     $9.25 
664                             Steak Salad     $8.99 
281                        Steak Salad Bowl    $23.78 
8                          Steak Soft Tacos     $9.25 
62                              Veggie Bowl    $11.25 
57                           Veggie Burrito    $11.25 
1653                    Veggie Crispy Tacos     $8.49 
1694                           Veggie Salad     $8.49 
186                       Veggie Salad Bowl    $11.25 
738                       Veggie Soft Tacos    $11.25

## What was the quantity of the most expensive item ordered?

In [103]:
#gotta remove the $ sign from the item_price column:
price=[ float(ds1['item_price'][i][1:]) for i in range(len(ds1)) ]
price=pd.DataFrame(price)

#assigning
ds1[['item_price']] = price
ds1.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  item_price Amount_paid  \
0                                                NaN        2.39       $2.39   
1                                       [Clementine]        3.39       $3.39   
2                                            [Apple]        3.39       $3.39   
3                                                NaN        2.39       $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98      $33.96   

                           items_offered  No. of times an item is bought  
0           Chips and Fresh Tomato Salsa                           130.0  
1                                   Izze                            20.0  
2                       Nantucket Nectar                            29.0  
3  Chips and Tomatillo-Green Chili Salsa                            33.0  
4                           Chicken Bowl                           761.0

In [108]:
ds1.iloc[ds1[ds1['item_price'] == max(ds1['item_price'])].index]

order_id  quantity                     item_name choice_description  \
3598      1443        15  Chips and Fresh Tomato Salsa                NaN   

      item_price Amount_paid items_offered  No. of times an item is bought  
3598       44.25     $663.75           NaN                             NaN

**quantity = 15**

## How many times were a Veggie Salad Bowl ordered?

In [148]:
ds1[ds1['item_name'] == 'Veggie Salad']

order_id  quantity     item_name  \
1694       686         1  Veggie Salad   
2756      1094         1  Veggie Salad   
2996      1192         1  Veggie Salad   
3163      1263         1  Veggie Salad   
3471      1395         1  Veggie Salad   
4084      1635         1  Veggie Salad   

                                     choice_description  item_price  \
1694  [[Fresh Tomato Salsa (Mild), Roasted Chili Cor...        8.49   
2756  [[Tomatillo-Green Chili Salsa (Medium), Roaste...        8.49   
2996  [Roasted Chili Corn Salsa (Medium), [Black Bea...        8.49   
3163  [[Fresh Tomato Salsa (Mild), Roasted Chili Cor...        8.49   
3471  [[Tomatillo-Green Chili Salsa (Medium), Roaste...        8.49   
4084  [[Fresh Tomato Salsa (Mild), Roasted Chili Cor...        8.49   

     Amount_paid items_offered  No. of times an item is bought  
1694       $8.49           NaN                             NaN  
2756       $8.49           NaN                             NaN  
2996       $8.49           NaN                             NaN  
3163       $8.49           NaN                             NaN  
3471       $8.49           NaN                             NaN  
4084       $8.49           NaN                             NaN

In [149]:
print(f"Veggie Salad was ordered {len(ds1[ds1['item_name'] == 'Veggie Salad'])} times.")

Veggie Salad was ordered 6 times.


In [18]:
#Another Approach:
ds1.groupby("item_name").sum().loc[['Veggie Salad']]

order_id  quantity
item_name                       
Veggie Salad      7265         6

## DataSet - 2: Beer Content

In [4]:
ds2=pd.read_table("https://raw.githubusercontent.com/justmarkham/DAT8/master/data/beer.txt", sep=" ")
ds2

name  calories  sodium  alcohol  cost
0              Budweiser       144      15      4.7  0.43
1                Schlitz       151      19      4.9  0.43
2              Lowenbrau       157      15      0.9  0.48
3            Kronenbourg       170       7      5.2  0.73
4               Heineken       152      11      5.0  0.77
5          Old_Milwaukee       145      23      4.6  0.28
6             Augsberger       175      24      5.5  0.40
7   Srohs_Bohemian_Style       149      27      4.7  0.42
8            Miller_Lite        99      10      4.3  0.43
9        Budweiser_Light       113       8      3.7  0.40
10                 Coors       140      18      4.6  0.44
11           Coors_Light       102      15      4.1  0.46
12        Michelob_Light       135      11      4.2  0.50
13                 Becks       150      19      4.7  0.76
14                 Kirin       149       6      5.0  0.79
15     Pabst_Extra_Light        68      15      2.3  0.38
16                 Hamms       139      19      4.4  0.43
17   Heilemans_Old_Style       144      24      4.9  0.43
18   Olympia_Goled_Light        72       6      2.9  0.46
19         Schlitz_Light        97       7      4.2  0.47

In [5]:
ds2.rename(columns={'name': 'Beer'}, inplace = True) 
#inlace: Whether to return a new DataFrame. 
#If True then value of copy is ignored.
ds2

Beer  calories  sodium  alcohol  cost
0              Budweiser       144      15      4.7  0.43
1                Schlitz       151      19      4.9  0.43
2              Lowenbrau       157      15      0.9  0.48
3            Kronenbourg       170       7      5.2  0.73
4               Heineken       152      11      5.0  0.77
5          Old_Milwaukee       145      23      4.6  0.28
6             Augsberger       175      24      5.5  0.40
7   Srohs_Bohemian_Style       149      27      4.7  0.42
8            Miller_Lite        99      10      4.3  0.43
9        Budweiser_Light       113       8      3.7  0.40
10                 Coors       140      18      4.6  0.44
11           Coors_Light       102      15      4.1  0.46
12        Michelob_Light       135      11      4.2  0.50
13                 Becks       150      19      4.7  0.76
14                 Kirin       149       6      5.0  0.79
15     Pabst_Extra_Light        68      15      2.3  0.38
16                 Hamms       139      19      4.4  0.43
17   Heilemans_Old_Style       144      24      4.9  0.43
18   Olympia_Goled_Light        72       6      2.9  0.46
19         Schlitz_Light        97       7      4.2  0.47

In [6]:
ds2.describe()

calories     sodium    alcohol       cost
count   20.000000  20.000000  20.000000  20.000000
mean   132.550000  14.950000   4.240000   0.494500
std     30.283702   6.581273   1.087924   0.144858
min     68.000000   6.000000   0.900000   0.280000
25%    110.250000   9.500000   4.175000   0.427500
50%    144.000000  15.000000   4.600000   0.435000
75%    150.250000  19.000000   4.900000   0.485000
max    175.000000  27.000000   5.500000   0.790000

In [7]:
pd.Categorical(ds2['Beer'])

['Budweiser', 'Schlitz', 'Lowenbrau', 'Kronenbourg', 'Heineken', ..., 'Pabst_Extra_Light', 'Hamms', 'Heilemans_Old_Style', 'Olympia_Goled_Light', 'Schlitz_Light']
Length: 20
Categories (20, object): ['Augsberger', 'Becks', 'Budweiser', 'Budweiser_Light', ..., 'Pabst_Extra_Light', 'Schlitz', 'Schlitz_Light', 'Srohs_Bohemian_Style']

In [8]:
ds2[['Beer']].describe()

Beer
count          20
unique         20
top     Budweiser
freq            1

### Beer w highest content of:
 - ### calories
 - ### sodium 
 - ### alcohol

In [8]:
ds2[ds2['calories'] == max(ds2['calories'])]

name  calories  sodium  alcohol  cost
6  Augsberger       175      24      5.5   0.4

**Augsberger** is the one with **highest calorie count.**

In [12]:
ds2[ds2['sodium'] == max(ds2['sodium'])]

name  calories  sodium  alcohol  cost
7  Srohs_Bohemian_Style       149      27      4.7  0.42

**Srohs_Bohemian_Style** is the one with **highest sodium count.**

In [13]:
ds2[ds2['alcohol'] == max(ds2['alcohol'])]

name  calories  sodium  alcohol  cost
6  Augsberger       175      24      5.5   0.4

**Augsberger** is also the one with **highest alcohol count.** thus, **Strongest Beer.**

### Most Expensive Beer:

In [19]:
ds2[ds2['cost'] == max(ds2['cost'])]

name  calories  sodium  alcohol  cost
14  Kirin       149       6      5.0  0.79

**Kirin** is the **most expensive beer.**

In [53]:
ds2.head(10)

Beer  calories  sodium  alcohol  cost
0             Budweiser       144      15      4.7  0.43
1               Schlitz       151      19      4.9  0.43
2             Lowenbrau       157      15      0.9  0.48
3           Kronenbourg       170       7      5.2  0.73
4              Heineken       152      11      5.0  0.77
5         Old_Milwaukee       145      23      4.6  0.28
6            Augsberger       175      24      5.5  0.40
7  Srohs_Bohemian_Style       149      27      4.7  0.42
8           Miller_Lite        99      10      4.3  0.43
9       Budweiser_Light       113       8      3.7  0.40

### Most inexpensive Beer:

In [57]:
ds2.groupby(ds2['Beer']).sum().sort_values(['cost']).head(1)

calories  sodium  alcohol  cost
Beer                                          
Old_Milwaukee       145      23      4.6  0.28

# DataSet 3: US Crime Rates (1960-2014)

In [34]:
crime=pd.read_csv("https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/04_Apply/US_Crime_Rates/US_Crime_Rates_1960_2014.csv")
crime.head(10)

Year  Population    Total  Violent  Property  Murder  Forcible_Rape  \
0  1960   179323175  3384200   288460   3095700    9110          17190   
1  1961   182992000  3488000   289390   3198600    8740          17220   
2  1962   185771000  3752200   301510   3450700    8530          17550   
3  1963   188483000  4109500   316970   3792500    8640          17650   
4  1964   191141000  4564600   364220   4200400    9360          21420   
5  1965   193526000  4739400   387390   4352000    9960          23410   
6  1966   195576000  5223500   430180   4793300   11040          25820   
7  1967   197457000  5903400   499930   5403500   12240          27620   
8  1968   199399000  6720200   595010   6125200   13800          31670   
9  1969   201385000  7410900   661870   6749000   14760          37170   

   Robbery  Aggravated_assault  Burglary  Larceny_Theft  Vehicle_Theft  
0   107840              154320    912100        1855400         328200  
1   106670              156760    949600        1913000         336000  
2   110860              164570    994300        2089600         366800  
3   116470              174210   1086400        2297800         408300  
4   130390              203050   1213200        2514400         472800  
5   138690              215330   1282500        2572600         496900  
6   157990              235330   1410100        2822000         561200  
7   202910              257160   1632100        3111600         659800  
8   262840              286700   1858900        3482700         783600  
9   298850              311090   1981900        3888600         878500

### Convert the type of the column Year to datetime64

In [35]:
crime.dtypes

Year                  int64
Population            int64
Total                 int64
Violent               int64
Property              int64
Murder                int64
Forcible_Rape         int64
Robbery               int64
Aggravated_assault    int64
Burglary              int64
Larceny_Theft         int64
Vehicle_Theft         int64
dtype: object

In [36]:
crime.Vehicle_Theft.head()
#or
crime['Vehicle_Theft'].head()

0    328200
1    336000
2    366800
3    408300
4    472800
Name: Vehicle_Theft, dtype: int64

In [37]:
crime['Year'] = pd.to_datetime(crime['Year'], format='%Y') #Typecasting
crime.info()
crime.head()
crime

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55 entries, 0 to 54
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Year                55 non-null     datetime64[ns]
 1   Population          55 non-null     int64         
 2   Total               55 non-null     int64         
 3   Violent             55 non-null     int64         
 4   Property            55 non-null     int64         
 5   Murder              55 non-null     int64         
 6   Forcible_Rape       55 non-null     int64         
 7   Robbery             55 non-null     int64         
 8   Aggravated_assault  55 non-null     int64         
 9   Burglary            55 non-null     int64         
 10  Larceny_Theft       55 non-null     int64         
 11  Vehicle_Theft       55 non-null     int64         
dtypes: datetime64[ns](1), int64(11)
memory usage: 5.3 KB


Year  Population     Total  Violent  Property  Murder  Forcible_Rape  \
0  1960-01-01   179323175   3384200   288460   3095700    9110          17190   
1  1961-01-01   182992000   3488000   289390   3198600    8740          17220   
2  1962-01-01   185771000   3752200   301510   3450700    8530          17550   
3  1963-01-01   188483000   4109500   316970   3792500    8640          17650   
4  1964-01-01   191141000   4564600   364220   4200400    9360          21420   
5  1965-01-01   193526000   4739400   387390   4352000    9960          23410   
6  1966-01-01   195576000   5223500   430180   4793300   11040          25820   
7  1967-01-01   197457000   5903400   499930   5403500   12240          27620   
8  1968-01-01   199399000   6720200   595010   6125200   13800          31670   
9  1969-01-01   201385000   7410900   661870   6749000   14760          37170   
10 1970-01-01   203235298   8098000   738820   7359200   16000          37990   
11 1971-01-01   206212000   8588200   816500   7771700   17780          42260   
12 1972-01-01   208230000   8248800   834900   7413900   18670          46850   
13 1973-01-01   209851000   8718100   875910   7842200   19640          51400   
14 1974-01-01   211392000  10253400   974720   9278700   20710          55400   
15 1975-01-01   213124000  11292400  1039710  10252700   20510          56090   
16 1976-01-01   214659000  11349700  1004210  10345500   18780          57080   
17 1977-01-01   216332000  10984500  1029580   9955000   19120          63500   
18 1978-01-01   218059000  11209000  1085550  10123400   19560          67610   
19 1979-01-01   220099000  12249500  1208030  11041500   21460          76390   
20 1980-01-01   225349264  13408300  1344520  12063700   23040          82990   
21 1981-01-01   229146000  13423800  1361820  12061900   22520          82500   
22 1982-01-01   231534000  12974400  1322390  11652000   21010          78770   
23 1983-01-01   233981000  12108600  1258090  10850500   19310          78920   
24 1984-01-01   236158000  11881800  1273280  10608500   18690          84230   
25 1985-01-01   238740000  12431400  1328800  11102600   18980          88670   
26 1986-01-01   240132887  13211869  1489169  11722700   20613          91459   
27 1987-01-01   242282918  13508700  1483999  12024700   20096          91110   
28 1988-01-01   245807000  13923100  1566220  12356900   20680          92490   
29 1989-01-01   248239000  14251400  1646040  12605400   21500          94500   
30 1990-01-01   248709873  14475600  1820130  12655500   23440         102560   
31 1991-01-01   252177000  14872900  1911770  12961100   24700         106590   
32 1992-01-01   255082000  14438200  1932270  12505900   23760         109060   
33 1993-01-01   257908000  14144800  1926020  12218800   24530         106010   
34 1994-01-01   260341000  13989500  1857670  12131900   23330         102220   
35 1995-01-01   262755000  13862700  1798790  12063900   21610          97470   
36 1996-01-01   265228572  13493863  1688540  11805300   19650          96250   
37 1997-01-01   267637000  13194571  1634770  11558175   18208          96153   
38 1998-01-01   270296000  12475634  1531044  10944590   16914          93103   
39 1999-01-01   272690813  11634378  1426044  10208334   15522          89411   
40 2000-01-01   281421906  11608072  1425486  10182586   15586          90178   
41 2001-01-01   285317559  11876669  1439480  10437480   16037          90863   
42 2002-01-01   287973924  11878954  1423677  10455277   16229          95235   
43 2003-01-01   290690788  11826538  1383676  10442862   16528          93883   
44 2004-01-01   293656842  11679474  1360088  10319386   16148          95089   
45 2005-01-01   296507061  11565499  1390745  10174754   16740          94347   
46 2006-01-01   299398484  11401511  1418043   9983568   17030          92757   
47 2007-01-01   301621157  11251828  1408337   9843481   16929          90427   
48 2008-01-01   304374846  11160543  139

### Set the Year column as the index of the dataframe

In [38]:
crime.index = crime.Year
crime = crime.drop(['Year'], axis=1)
crime.head()

Population    Total  Violent  Property  Murder  Forcible_Rape  \
Year                                                                        
1960-01-01   179323175  3384200   288460   3095700    9110          17190   
1961-01-01   182992000  3488000   289390   3198600    8740          17220   
1962-01-01   185771000  3752200   301510   3450700    8530          17550   
1963-01-01   188483000  4109500   316970   3792500    8640          17650   
1964-01-01   191141000  4564600   364220   4200400    9360          21420   

            Robbery  Aggravated_assault  Burglary  Larceny_Theft  \
Year                                                               
1960-01-01   107840              154320    912100        1855400   
1961-01-01   106670              156760    949600        1913000   
1962-01-01   110860              164570    994300        2089600   
1963-01-01   116470              174210   1086400        2297800   
1964-01-01   130390              203050   1213200        2514400   

            Vehicle_Theft  
Year                       
1960-01-01         328200  
1961-01-01         336000  
1962-01-01         366800  
1963-01-01         408300  
1964-01-01         472800

In [39]:
# OR
try:
    crime = crime.set_index('Year', drop = True)
except Exception as e:
    print(e)
crime.head()

"None of ['Year'] are in the columns"


Population    Total  Violent  Property  Murder  Forcible_Rape  \
Year                                                                        
1960-01-01   179323175  3384200   288460   3095700    9110          17190   
1961-01-01   182992000  3488000   289390   3198600    8740          17220   
1962-01-01   185771000  3752200   301510   3450700    8530          17550   
1963-01-01   188483000  4109500   316970   3792500    8640          17650   
1964-01-01   191141000  4564600   364220   4200400    9360          21420   

            Robbery  Aggravated_assault  Burglary  Larceny_Theft  \
Year                                                               
1960-01-01   107840              154320    912100        1855400   
1961-01-01   106670              156760    949600        1913000   
1962-01-01   110860              164570    994300        2089600   
1963-01-01   116470              174210   1086400        2297800   
1964-01-01   130390              203050   1213200        2514400   

            Vehicle_Theft  
Year                       
1960-01-01         328200  
1961-01-01         336000  
1962-01-01         366800  
1963-01-01         408300  
1964-01-01         472800

### Group the Year by decades and sum the values accordingly:

In [29]:
crime[::10] 

Population     Total  Violent  Property  Murder  Forcible_Rape  Robbery  \
Year                                                                            
1960   179323175   3384200   288460   3095700    9110          17190   107840   
1970   203235298   8098000   738820   7359200   16000          37990   349860   
1980   225349264  13408300  1344520  12063700   23040          82990   565840   
1990   248709873  14475600  1820130  12655500   23440         102560   639270   
2000   281421906  11608072  1425486  10182586   15586          90178   408016   
2010   309330219  10363873  1251248   9112625   14772          85593   369089   

      Aggravated_assault  Burglary  Larceny_Theft  Vehicle_Theft  
Year                                                              
1960              154320    912100        1855400         328200  
1970              334970   2205000        4225800         928400  
1980              672650   3795200        7136900        1131700  
1990             1054860   3073900        7945700        1635900  
2000              911706   2050992        6971590        1160002  
2010              781844   2168457        6204601         739565

In [40]:
#using resample to resample the data
crimes=crime.resample('10AS').sum()

#We oughta take the max population of the decade rather than the sum
#pretty much makes sense
pop = crime['Population'].resample('10AS').max()

#assigning the max population column to our new df crimes.
crimes['Population'] = pop

crimes

Population      Total   Violent   Property  Murder  Forcible_Rape  \
Year                                                                            
1960-01-01   201385000   49295900   4134930   45160900  106180         236720   
1970-01-01   220099000  100991600   9607930   91383800  192230         554570   
1980-01-01   248239000  131123369  14074328  117048900  206439         865639   
1990-01-01   272690813  136582146  17527048  119053499  211664         998827   
2000-01-01   307006550  115012044  13968056  100944369  163068         922499   
2010-01-01   318857056   50167967   6072017   44095950   72867         421059   

            Robbery  Aggravated_assault  Burglary  Larceny_Theft  \
Year                                                               
1960-01-01  1633510             2158520  13321100       26547700   
1970-01-01  4159020             4702120  28486000       53157800   
1980-01-01  5383109             7619130  33073494       72040253   
1990-01-01  5748930            10568963  26750015       77679366   
2000-01-01  4230366             8652124  21565176       67970291   
2010-01-01  1749809             3764142  10125170       30401698   

            Vehicle_Theft  
Year                       
1960-01-01        5292100  
1970-01-01        9739900  
1980-01-01       11935411  
1990-01-01       14624418  
2000-01-01       11412834  
2010-01-01        3569080

### Return the first 3 rows of the this DataSet.

### When it comes to selecting rows and columns of a pandas DataFrame, loc and iloc are two commonly used functions.<br>Here is the subtle difference between the two functions: loc selects rows and columns with specific labels. iloc selects rows and columns at specific integer positions.

In [103]:
crime.head(3)
#or
crime.iloc[range(3)]

Population    Total  Violent  Property  Murder  Forcible_Rape  \
Year                                                                        
1960-01-01   179323175  3384200   288460   3095700    9110          17190   
1961-01-01   182992000  3488000   289390   3198600    8740          17220   
1962-01-01   185771000  3752200   301510   3450700    8530          17550   

            Robbery  Aggravated_assault  Burglary  Larceny_Theft  \
Year                                                               
1960-01-01   107840              154320    912100        1855400   
1961-01-01   106670              156760    949600        1913000   
1962-01-01   110860              164570    994300        2089600   

            Vehicle_Theft  
Year                       
1960-01-01         328200  
1961-01-01         336000  
1962-01-01         366800

### Select just the 'Murder' and 'Robbery' columns from the DataFrame crime

In [93]:
crime[['Murder', 'Robbery']].head()

Murder  Robbery
Year                       
1960-01-01    9110   107840
1961-01-01    8740   106670
1962-01-01    8530   110860
1963-01-01    8640   116470
1964-01-01    9360   130390

In [109]:
crime.loc[:, ['Murder', 'Robbery']].head()

Murder  Robbery
Year                       
1960-01-01    9110   107840
1961-01-01    8740   106670
1962-01-01    8530   110860
1963-01-01    8640   116470
1964-01-01    9360   130390

### Select the data in rows [3, 4, 8] and in columns ['Murder', 'Robbery']

In [112]:
crime[['Murder', 'Robbery']].iloc[[3, 4, 8]]

Murder  Robbery
Year                       
1963-01-01    8640   116470
1964-01-01    9360   130390
1968-01-01   13800   262840

### Select the rows the murder is between 20k and 24k (inclusive)

In [137]:
req=crime[crime['Murder'] >= 20000]

req[req['Murder']<=24000]

Population     Total  Violent  Property  Murder  Forcible_Rape  \
Year                                                                         
1974-01-01   211392000  10253400   974720   9278700   20710          55400   
1975-01-01   213124000  11292400  1039710  10252700   20510          56090   
1979-01-01   220099000  12249500  1208030  11041500   21460          76390   
1980-01-01   225349264  13408300  1344520  12063700   23040          82990   
1981-01-01   229146000  13423800  1361820  12061900   22520          82500   
1982-01-01   231534000  12974400  1322390  11652000   21010          78770   
1986-01-01   240132887  13211869  1489169  11722700   20613          91459   
1987-01-01   242282918  13508700  1483999  12024700   20096          91110   
1988-01-01   245807000  13923100  1566220  12356900   20680          92490   
1989-01-01   248239000  14251400  1646040  12605400   21500          94500   
1990-01-01   248709873  14475600  1820130  12655500   23440         102560   
1992-01-01   255082000  14438200  1932270  12505900   23760         109060   
1994-01-01   260341000  13989500  1857670  12131900   23330         102220   
1995-01-01   262755000  13862700  1798790  12063900   21610          97470   

            Robbery  Aggravated_assault  Burglary  Larceny_Theft  \
Year                                                               
1974-01-01   442400              456210   3039200        5262500   
1975-01-01   470500              492620   3265300        5977700   
1979-01-01   480700              629480   3327700        6601000   
1980-01-01   565840              672650   3795200        7136900   
1981-01-01   592910              663900   3779700        7194400   
1982-01-01   553130              669480   3447100        7142500   
1986-01-01   542775              834322   3241410        7257153   
1987-01-01   517704              855088   3236184        7499900   
1988-01-01   542970              910090   3218100        7705900   
1989-01-01   578330              951710   3168200        7872400   
1990-01-01   639270             1054860   3073900        7945700   
1992-01-01   672480             1126970   2979900        7915200   
1994-01-01   618950             1113180   2712800        7879800   
1995-01-01   580510             1099210   2593800        7997700   

            Vehicle_Theft  
Year                       
1974-01-01         977100  
1975-01-01        1009600  
1979-01-01        1112800  
1980-01-01        1131700  
1981-01-01        1087800  
1982-01-01        1062400  
1986-01-01        1224137  
1987-01-01        1288674  
1988-01-01        1432900  
1989-01-01        1564800  
1990-01-01        1635900  
1992-01-01        1610800  
1994-01-01        1539300  
1995-01-01        1472400

### OR:

In [138]:
crime[crime['Murder'].between(20000, 24000)]

Population     Total  Violent  Property  Murder  Forcible_Rape  \
Year                                                                         
1974-01-01   211392000  10253400   974720   9278700   20710          55400   
1975-01-01   213124000  11292400  1039710  10252700   20510          56090   
1979-01-01   220099000  12249500  1208030  11041500   21460          76390   
1980-01-01   225349264  13408300  1344520  12063700   23040          82990   
1981-01-01   229146000  13423800  1361820  12061900   22520          82500   
1982-01-01   231534000  12974400  1322390  11652000   21010          78770   
1986-01-01   240132887  13211869  1489169  11722700   20613          91459   
1987-01-01   242282918  13508700  1483999  12024700   20096          91110   
1988-01-01   245807000  13923100  1566220  12356900   20680          92490   
1989-01-01   248239000  14251400  1646040  12605400   21500          94500   
1990-01-01   248709873  14475600  1820130  12655500   23440         102560   
1992-01-01   255082000  14438200  1932270  12505900   23760         109060   
1994-01-01   260341000  13989500  1857670  12131900   23330         102220   
1995-01-01   262755000  13862700  1798790  12063900   21610          97470   

            Robbery  Aggravated_assault  Burglary  Larceny_Theft  \
Year                                                               
1974-01-01   442400              456210   3039200        5262500   
1975-01-01   470500              492620   3265300        5977700   
1979-01-01   480700              629480   3327700        6601000   
1980-01-01   565840              672650   3795200        7136900   
1981-01-01   592910              663900   3779700        7194400   
1982-01-01   553130              669480   3447100        7142500   
1986-01-01   542775              834322   3241410        7257153   
1987-01-01   517704              855088   3236184        7499900   
1988-01-01   542970              910090   3218100        7705900   
1989-01-01   578330              951710   3168200        7872400   
1990-01-01   639270             1054860   3073900        7945700   
1992-01-01   672480             1126970   2979900        7915200   
1994-01-01   618950             1113180   2712800        7879800   
1995-01-01   580510             1099210   2593800        7997700   

            Vehicle_Theft  
Year                       
1974-01-01         977100  
1975-01-01        1009600  
1979-01-01        1112800  
1980-01-01        1131700  
1981-01-01        1087800  
1982-01-01        1062400  
1986-01-01        1224137  
1987-01-01        1288674  
1988-01-01        1432900  
1989-01-01        1564800  
1990-01-01        1635900  
1992-01-01        1610800  
1994-01-01        1539300  
1995-01-01        1472400

### OR:

In [41]:
crime.query('Murder >= 20000' and 'Murder <= 24000')

Population     Total  Violent  Property  Murder  Forcible_Rape  \
Year                                                                         
1960-01-01   179323175   3384200   288460   3095700    9110          17190   
1961-01-01   182992000   3488000   289390   3198600    8740          17220   
1962-01-01   185771000   3752200   301510   3450700    8530          17550   
1963-01-01   188483000   4109500   316970   3792500    8640          17650   
1964-01-01   191141000   4564600   364220   4200400    9360          21420   
1965-01-01   193526000   4739400   387390   4352000    9960          23410   
1966-01-01   195576000   5223500   430180   4793300   11040          25820   
1967-01-01   197457000   5903400   499930   5403500   12240          27620   
1968-01-01   199399000   6720200   595010   6125200   13800          31670   
1969-01-01   201385000   7410900   661870   6749000   14760          37170   
1970-01-01   203235298   8098000   738820   7359200   16000          37990   
1971-01-01   206212000   8588200   816500   7771700   17780          42260   
1972-01-01   208230000   8248800   834900   7413900   18670          46850   
1973-01-01   209851000   8718100   875910   7842200   19640          51400   
1974-01-01   211392000  10253400   974720   9278700   20710          55400   
1975-01-01   213124000  11292400  1039710  10252700   20510          56090   
1976-01-01   214659000  11349700  1004210  10345500   18780          57080   
1977-01-01   216332000  10984500  1029580   9955000   19120          63500   
1978-01-01   218059000  11209000  1085550  10123400   19560          67610   
1979-01-01   220099000  12249500  1208030  11041500   21460          76390   
1980-01-01   225349264  13408300  1344520  12063700   23040          82990   
1981-01-01   229146000  13423800  1361820  12061900   22520          82500   
1982-01-01   231534000  12974400  1322390  11652000   21010          78770   
1983-01-01   233981000  12108600  1258090  10850500   19310          78920   
1984-01-01   236158000  11881800  1273280  10608500   18690          84230   
1985-01-01   238740000  12431400  1328800  11102600   18980          88670   
1986-01-01   240132887  13211869  1489169  11722700   20613          91459   
1987-01-01   242282918  13508700  1483999  12024700   20096          91110   
1988-01-01   245807000  13923100  1566220  12356900   20680          92490   
1989-01-01   248239000  14251400  1646040  12605400   21500          94500   
1990-01-01   248709873  14475600  1820130  12655500   23440         102560   
1992-01-01   255082000  14438200  1932270  12505900   23760         109060   
1994-01-01   260341000  13989500  1857670  12131900   23330         102220   
1995-01-01   262755000  13862700  1798790  12063900   21610          97470   
1996-01-01   265228572  13493863  1688540  11805300   19650          96250   
1997-01-01   267637000  13194571  1634770  11558175   18208          96153   
1998-01-01   270296000  12475634  1531044  10944590   16914          93103   
1999-01-01   272690813  11634378  1426044  10208334   15522          89411   
2000-01-01   281421906  11608072  1425486  10182586   15586          90178   
2001-01-01   285317559  11876669  1439480  10437480   16037          90863   
2002-01-01   287973924  11878954  1423677  10455277   16229          95235   
2003-01-01   290690788  11826538  1383676  10442862   16528          93883   
2004-01-01   293656842  11679474  1360088  10319386   16148          95089   
2005-01-01   296507061  11565499  1390745  10174754   16740          94347   
2006-01-01   299398484  11401511  1418043   9983568   17030          92757   
2007-01-01   301621157  11251828  1408337   9843481   16929          90427   
2008-01-01   304374846  11160543  1392628   9767915   16442          90479   
2009-01-01   307006550  10762956  1325896   9337060   15399          89241   
2010-01-01   309330219  10363873  1251248   9112625   14772          85593   
2011-01-01   311587816  10258774  

### Calculate the mean murder for each different year in crime DataFrame.

In [143]:
crime.head()

Population    Total  Violent  Property  Murder  Forcible_Rape  \
Year                                                                        
1960-01-01   179323175  3384200   288460   3095700    9110          17190   
1961-01-01   182992000  3488000   289390   3198600    8740          17220   
1962-01-01   185771000  3752200   301510   3450700    8530          17550   
1963-01-01   188483000  4109500   316970   3792500    8640          17650   
1964-01-01   191141000  4564600   364220   4200400    9360          21420   

            Robbery  Aggravated_assault  Burglary  Larceny_Theft  \
Year                                                               
1960-01-01   107840              154320    912100        1855400   
1961-01-01   106670              156760    949600        1913000   
1962-01-01   110860              164570    994300        2089600   
1963-01-01   116470              174210   1086400        2297800   
1964-01-01   130390              203050   1213200        2514400   

            Vehicle_Theft  
Year                       
1960-01-01         328200  
1961-01-01         336000  
1962-01-01         366800  
1963-01-01         408300  
1964-01-01         472800

In [145]:
crime.groupby('Year')['Murder'].mean()

Year
1960-01-01     9110.0
1961-01-01     8740.0
1962-01-01     8530.0
1963-01-01     8640.0
1964-01-01     9360.0
1965-01-01     9960.0
1966-01-01    11040.0
1967-01-01    12240.0
1968-01-01    13800.0
1969-01-01    14760.0
1970-01-01    16000.0
1971-01-01    17780.0
1972-01-01    18670.0
1973-01-01    19640.0
1974-01-01    20710.0
1975-01-01    20510.0
1976-01-01    18780.0
1977-01-01    19120.0
1978-01-01    19560.0
1979-01-01    21460.0
1980-01-01    23040.0
1981-01-01    22520.0
1982-01-01    21010.0
1983-01-01    19310.0
1984-01-01    18690.0
1985-01-01    18980.0
1986-01-01    20613.0
1987-01-01    20096.0
1988-01-01    20680.0
1989-01-01    21500.0
1990-01-01    23440.0
1991-01-01    24700.0
1992-01-01    23760.0
1993-01-01    24530.0
1994-01-01    23330.0
1995-01-01    21610.0
1996-01-01    19650.0
1997-01-01    18208.0
1998-01-01    16914.0
1999-01-01    15522.0
2000-01-01    15586.0
2001-01-01    16037.0
2002-01-01    16229.0
2003-01-01    16528.0
2004-01-01    16148.0
2005-

### Sort df crime first by the values in the 'Murder' in descending order, then by the value in the 'Violent' column in ascending order.

In [149]:
crime[['Murder']].sort_values(['Murder'], ascending=False).head()

Murder
Year              
1991-01-01   24700
1993-01-01   24530
1992-01-01   23760
1990-01-01   23440
1994-01-01   23330

In [46]:
crime.sort_values(['Murder'], ascending=False).sort_values(['Violent'], ascending=True).head()

Population    Total  Violent  Property  Murder  Forcible_Rape  \
Year                                                                        
1960-01-01   179323175  3384200   288460   3095700    9110          17190   
1961-01-01   182992000  3488000   289390   3198600    8740          17220   
1962-01-01   185771000  3752200   301510   3450700    8530          17550   
1963-01-01   188483000  4109500   316970   3792500    8640          17650   
1964-01-01   191141000  4564600   364220   4200400    9360          21420   

            Robbery  Aggravated_assault  Burglary  Larceny_Theft  \
Year                                                               
1960-01-01   107840              154320    912100        1855400   
1961-01-01   106670              156760    949600        1913000   
1962-01-01   110860              164570    994300        2089600   
1963-01-01   116470              174210   1086400        2297800   
1964-01-01   130390              203050   1213200        2514400   

            Vehicle_Theft  
Year                       
1960-01-01         328200  
1961-01-01         336000  
1962-01-01         366800  
1963-01-01         408300  
1964-01-01         472800

In [152]:
crime.sort_values(['Murder','Violent'], ascending=[False, True]).head()

Population     Total  Violent  Property  Murder  Forcible_Rape  \
Year                                                                         
1991-01-01   252177000  14872900  1911770  12961100   24700         106590   
1993-01-01   257908000  14144800  1926020  12218800   24530         106010   
1992-01-01   255082000  14438200  1932270  12505900   23760         109060   
1990-01-01   248709873  14475600  1820130  12655500   23440         102560   
1994-01-01   260341000  13989500  1857670  12131900   23330         102220   

            Robbery  Aggravated_assault  Burglary  Larceny_Theft  \
Year                                                               
1991-01-01   687730             1092740   3157200        8142200   
1993-01-01   659870             1135610   2834800        7820900   
1992-01-01   672480             1126970   2979900        7915200   
1990-01-01   639270             1054860   3073900        7945700   
1994-01-01   618950             1113180   2712800        7879800   

            Vehicle_Theft  
Year                       
1991-01-01        1661700  
1993-01-01        1563100  
1992-01-01        1610800  
1990-01-01        1635900  
1994-01-01        1539300

# DataSet 4:

In [157]:
ds4 = pd.read_csv('https://raw.githubusercontent.com/vincentarelbundock/Rdatasets/master/datasets.csv')
ds4

Package           Item  \
0            AER        Affairs   
1            AER   ArgentinaCPI   
2            AER      BankWages   
3            AER  BenderlyZwick   
4            AER      BondYield   
...          ...            ...   
1879  wooldridge          wage1   
1880  wooldridge          wage2   
1881  wooldridge        wagepan   
1882  wooldridge        wageprc   
1883  wooldridge           wine   

                                                  Title  Rows  Cols  n_binary  \
0                      Fair's Extramarital Affairs Data   601     9         2   
1                     Consumer Price Index in Argentina    80     2         0   
2                                            Bank Wages   474     4         2   
3     Benderly and Zwick Data: Inflation, Growth and...    31     5         0   
4                                       Bond Yield Data    60     2         0   
...                                                 ...   ...   ...       ...   
1879                                              wage1   526    24        16   
1880                                              wage2   935    17         4   
1881                                            wagepan  4360    44        37   
1882                                            wageprc   286    20         0   
1883                                               wine    21     5         0   

      n_character  n_factor  n_logical  n_numeric  \
0               0         2          0          7   
1               0         0          0          2   
2               0         3          0          1   
3               0         0          0          5   
4               0         0          0          2   
...           ...       ...        ...        ...   
1879            0         0          0         24   
1880            0         0          0         17   
1881            0         0          0         44   
1882            0         0          0         20   
1883            1         0          0          4   

                                                    CSV  \
0     https://vincentarelbundock.github.io/Rdatasets...   
1     https://vincentarelbundock.github.io/Rdatasets...   
2     https://vincentarelbundock.github.io/Rdatasets...   
3     https://vincentarelbundock.github.io/Rdatasets...   
4     https://vincentarelbundock.github.io/Rdatasets...   
...                                                 ...   
1879  https://vincentarelbundock.github.io/Rdatasets...   
1880  https://vincentarelbundock.github.io/Rdatasets...   
1881  https://vincentarelbundock.github.io/Rdatasets...   
1882  https://vincentarelbundock.github.io/Rdatasets...   
1883  https://vincentarelbundock.github.io/Rdatasets...   

                                                    Doc  
0     https://vincentarelbundock.github.io/Rdatasets...  
1     https://vincentarelbundock.github.io/Rdatasets...  
2     https://vincentarelbundock.github.io/Rdatasets...  
3     https://vincentarelbundock.github.io/Rdatasets...  
4     https://vincentarelbundock.github.io/Rdatasets...  
...                                                 ...  
1879  https://vincentarelbundock.github.io/Rdatasets...  
1880  https://vincentarelbundock.github.io/Rdatasets...  
1881  https://vincentarelbundock.github.io/Rdatasets...  
1882  https://vincentarelbundock.github.io/Rdatasets...  
1883  https://vincentarelbundock.github.io/Rdatasets...  

[1884 rows x 12 columns]

In [195]:
ds4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1884 entries, 0 to 1883
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Package      1884 non-null   object
 1   Item         1884 non-null   object
 2   Title        1884 non-null   object
 3   Rows         1884 non-null   int64 
 4   Cols         1884 non-null   int64 
 5   n_binary     1884 non-null   int64 
 6   n_character  1884 non-null   int64 
 7   n_factor     1884 non-null   int64 
 8   n_logical    1884 non-null   int64 
 9   n_numeric    1884 non-null   int64 
 10  CSV          1884 non-null   object
 11  Doc          1884 non-null   object
dtypes: int64(7), object(5)
memory usage: 176.8+ KB


In [201]:
ds4.shape #(rows, columns)

(1884, 12)

In [189]:
a=ds4[['Rows']].sum()
b=ds4[['Cols']].sum()
#Total no. of elements in whole tables contained in the rows of ds4
tot = int(a)*int(b)
tot

189963673326

In [165]:
ds4.describe()

Rows         Cols     n_binary  n_character     n_factor  \
count  1.884000e+03  1884.000000  1884.000000  1884.000000  1884.000000   
mean   3.955513e+03    13.530255     2.225053     0.351911     1.201168   
std    3.868900e+04   164.079096    16.781096     1.476976     3.108992   
min    2.000000e+00     1.000000     0.000000     0.000000     0.000000   
25%    3.800000e+01     3.000000     0.000000     0.000000     0.000000   
50%    1.210000e+02     5.000000     0.000000     0.000000     0.000000   
75%    6.960000e+02    10.000000     2.000000     0.000000     2.000000   
max    1.414593e+06  6831.000000   624.000000    29.000000    64.000000   

         n_logical    n_numeric  
count  1884.000000  1884.000000  
mean      0.028132    11.900743  
std       0.386653   164.021527  
min       0.000000     0.000000  
25%       0.000000     2.000000  
50%       0.000000     4.000000  
75%       0.000000     8.000000  
max      11.000000  6830.000000

In [204]:
ds4[['CSV']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1884 entries, 0 to 1883
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   CSV     1884 non-null   object
dtypes: object(1)
memory usage: 14.8+ KB


In [169]:
ds4[['Item']].describe()

Item
count       1884
unique      1822
top     salinity
freq           3

# DataSet 5:

In [175]:
ds5=pd.read_html("https://vincentarelbundock.github.io/Rdatasets/datasets.html")

In [181]:
len(ds5) #i.e two tables are there in the html page.

2

In [193]:
d50=ds5[0]
d50

Package           Item  \
0                                                   AER        Affairs   
1                                                   AER   ArgentinaCPI   
2                                                   AER      BankWages   
3                                                   AER  BenderlyZwick   
4                                                   AER      BondYield   
...                                                 ...            ...   
1880                                         wooldridge          wage2   
1881                                         wooldridge        wagepan   
1882                                         wooldridge        wageprc   
1883                                         wooldridge           wine   
1884  Package Item Title Rows Cols n_binary n_charac...            NaN   

                                                  Title    Rows  Cols  \
0                      Fair's Extramarital Affairs Data   601.0   9.0   
1                     Consumer Price Index in Argentina    80.0   2.0   
2                                            Bank Wages   474.0   4.0   
3     Benderly and Zwick Data: Inflation, Growth and...    31.0   5.0   
4                                       Bond Yield Data    60.0   2.0   
...                                                 ...     ...   ...   
1880                                              wage2   935.0  17.0   
1881                                            wagepan  4360.0  44.0   
1882                                            wageprc   286.0  20.0   
1883                                               wine    21.0   5.0   
1884                                                NaN     NaN   NaN   

      n_binary  n_character  n_factor  n_logical  n_numeric  CSV  Doc  
0          2.0          0.0       2.0        0.0        7.0  CSV  DOC  
1          0.0          0.0       0.0        0.0        2.0  CSV  DOC  
2          2.0          0.0       3.0        0.0        1.0  CSV  DOC  
3          0.0          0.0       0.0        0.0        5.0  CSV  DOC  
4          0.0          0.0       0.0        0.0        2.0  CSV  DOC  
...        ...          ...       ...        ...        ...  ...  ...  
1880       4.0          0.0       0.0        0.0       17.0  CSV  DOC  
1881      37.0          0.0       0.0        0.0       44.0  CSV  DOC  
1882       0.0          0.0       0.0        0.0       20.0  CSV  DOC  
1883       0.0          1.0       0.0        0.0        4.0  CSV  DOC  
1884       NaN          NaN       NaN        NaN        NaN  NaN  NaN  

[1885 rows x 12 columns]

In [194]:
d50.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1885 entries, 0 to 1884
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Package      1885 non-null   object 
 1   Item         1884 non-null   object 
 2   Title        1884 non-null   object 
 3   Rows         1884 non-null   float64
 4   Cols         1884 non-null   float64
 5   n_binary     1884 non-null   float64
 6   n_character  1884 non-null   float64
 7   n_factor     1884 non-null   float64
 8   n_logical    1884 non-null   float64
 9   n_numeric    1884 non-null   float64
 10  CSV          1884 non-null   object 
 11  Doc          1884 non-null   object 
dtypes: float64(7), object(5)
memory usage: 176.8+ KB


In [191]:
d51=ds5[1]
d51

Package           Item  \
0            AER        Affairs   
1            AER   ArgentinaCPI   
2            AER      BankWages   
3            AER  BenderlyZwick   
4            AER      BondYield   
...          ...            ...   
1879  wooldridge          wage1   
1880  wooldridge          wage2   
1881  wooldridge        wagepan   
1882  wooldridge        wageprc   
1883  wooldridge           wine   

                                                  Title  Rows  Cols  n_binary  \
0                      Fair's Extramarital Affairs Data   601     9         2   
1                     Consumer Price Index in Argentina    80     2         0   
2                                            Bank Wages   474     4         2   
3     Benderly and Zwick Data: Inflation, Growth and...    31     5         0   
4                                       Bond Yield Data    60     2         0   
...                                                 ...   ...   ...       ...   
1879                                              wage1   526    24        16   
1880                                              wage2   935    17         4   
1881                                            wagepan  4360    44        37   
1882                                            wageprc   286    20         0   
1883                                               wine    21     5         0   

      n_character  n_factor  n_logical  n_numeric  CSV  Doc  
0               0         2          0          7  CSV  DOC  
1               0         0          0          2  CSV  DOC  
2               0         3          0          1  CSV  DOC  
3               0         0          0          5  CSV  DOC  
4               0         0          0          2  CSV  DOC  
...           ...       ...        ...        ...  ...  ...  
1879            0         0          0         24  CSV  DOC  
1880            0         0          0         17  CSV  DOC  
1881            0         0          0         44  CSV  DOC  
1882            0         0          0         20  CSV  DOC  
1883            1         0          0          4  CSV  DOC  

[1884 rows x 12 columns]

In [192]:
d51.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1884 entries, 0 to 1883
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Package      1884 non-null   object
 1   Item         1884 non-null   object
 2   Title        1884 non-null   object
 3   Rows         1884 non-null   int64 
 4   Cols         1884 non-null   int64 
 5   n_binary     1884 non-null   int64 
 6   n_character  1884 non-null   int64 
 7   n_factor     1884 non-null   int64 
 8   n_logical    1884 non-null   int64 
 9   n_numeric    1884 non-null   int64 
 10  CSV          1884 non-null   object
 11  Doc          1884 non-null   object
dtypes: int64(7), object(5)
memory usage: 176.8+ KB
